keps/349-all-or-nothing/README.md

```diff
@@ -0,0 +1,466 @@
+block admission of new workloads until the first batch of the Job is scheduled.
+Additionally, we await with starting the Job until all other admitted workloads
+have their pods scheduled. This behavior is opted-in by jobs by being submitted
+to a localqueue configured to enable it.
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As an administrator of a batch processing in my organization I want to be able
+to configure kueue quotas for on-demand resources. However, I don't want to risk
+two Jobs deadlock in case of issues with provisioning nodes to satisfy the configured
+cluster queue quota. This could happen when the job corresponding to these Jobs don't
+specify priorities or specify the same priority.
+
+My use case can be supported by creating a dedicated localqueue with the
+`workloadStartupRequirements: QuotaAndPodsReady`
+annotation to all jobs. This local queue will be used by users to submit
+workloads which require all pods to be running at the same time.
+
+### Notes/Contraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+#### Choosing the API object to opt-in for the mechanism
+
+Let's review the options for the best place for the API and their pros&cons:
+- job
+- local queue
+- cluster queue
+- component
+
+**Workload/Job level pros**:
+- throughput - the same cluster queue may run workloads which don't need any guarantees and thus get admitted faster.
+- flexibility -  we can easily implement queue-level (or component-level) by translating the queue-level property into workload property during workload creation. For example, add the workload annotation to every workload created for a given local queue.
+- conceptually, requiring all pods to run at the same time is a property of a workload itself.
+
+**Workload/Job level cons**:
+- introducing a new annotation on a Job level will increase the complexity of the job reconciler which is a reference Job-based implementation of CRD <> Workload integration. Similarly, introducing
+  a timeout on reaching `PodsReady` condition could increase the complexity of the reference implementation.
```

Comment:
> introducing a new annotation on a Job level will increase the complexity of the job reconciler which is a reference Job-based implementation of CRD <> Workload integration.

I don't think it increases the complexity. The main complexity is in watching the job status and translating that to a condition on the workload, this will not go away for any of the options. The extra thing this option adds is copying an annotation to the workload field.

> Similarly, introducing
  a timeout on reaching `PodsReady` condition could increase the complexity of the reference implementation.

Why? the waiting should be done by Kueue on the Workload condition, the job controller should do nothing.

Reply:
> I don't think it increases the complexity. The main complexity is in watching the job status and translating that to a condition on the workload, this will not go away for any of the options. The extra thing this option adds is copying an annotation to the workload field.

> Why? the waiting should be done by Kueue on the Workload condition, the job controller should do nothing.

Indeed, just copying over the field, so maybe complexity is not the best word, but it would increase the contract on the JobReconciler to do it, and if we consider Job <> Workload as a reference contract, it means that some future Crd reconcilers also should do it.  

It probably also means easier upgrade  if we rename the annotation or delete - it would be a change in the yaml for the local queue, rather than many Job yaml files.

Aside from this argument there are two smaller ones:
- localqueue would be a natural place then to put also the timeout, so that we also increase flexibility by allowing multiple timeouts 
- validation - not sure we want a webhook for job to validate the annotation, but we already have one for LocalQueue. WDYT?

Reply:
I think the general con here is increasing the api surface of the job api via an annotation, and annotations generally shouldn't be used as APIs.

> but it would increase the contract on the JobReconciler to do it, and if we consider Job <> Workload as a reference contract, it means that some future Crd reconcilers also should do it.

We need to put things into perspective. The reconciler will need to do a lot more than that already to support the feature whether or not we have this field. In the v1.Job case we are lucky that we already have the ready status on the job, and so we only need to update the integration controller; other custom jobs may not have that and so to support this feature will require changes to their original controllers not just the integration one. My point is that the integration will not be transparent by simply hiding the API. In fact, it might lead to potential inconsistencies: users assume that they can use the feature by setting the localQueue field, but the integration controller doesn't implement populating the Workload's PodsReady condition status.

> It probably also means easier upgrade if we rename the annotation or delete - it would be a change in the yaml for the local queue, rather than many Job yaml files.

I don't think renaming is something we need to be concerned about. It is more the deprecation that we need to ensure we can handle properly.

> localqueue would be a natural place then to put also the timeout, so that we also increase flexibility by allowing multiple timeouts

We discussed that the timeout is probably best set as a global parameter, I wouldn't put that in any of the queues APIs.

> validation - not sure we want a webhook for job to validate the annotation, but we already have one for LocalQueue. WDYT?

We also have a webhook for job to set suspend=true.

Reply:
Got it, reverting to the Job annotation idea, but going to describe the defaulting at the LocalQueue level as an alternative for potential further discussion.